cpp/src/skyhook/protocol/skyhook_protocol.cc

```diff
@@ -0,0 +1,139 @@
+      arrow::ipc::SerializeSchema(*req.projection_schema).ValueOrDie();
+  auto dataset_schema = arrow::ipc::SerializeSchema(*req.dataset_schema).ValueOrDie();
+
+  flatbuffers::FlatBufferBuilder builder(1024);
+  auto filter_expression_vector =
+      builder.CreateVector(filter_expression->data(), filter_expression->size());
+  auto partition_expression_vector =
+      builder.CreateVector(partition_expression->data(), partition_expression->size());
+  auto projected_schema_vector =
+      builder.CreateVector(projection_schema->data(), projection_schema->size());
+  auto dataset_schema_vector =
+      builder.CreateVector(dataset_schema->data(), dataset_schema->size());
+
+  auto request = flatbuf::CreateScanRequest(
+      builder, req.file_size, static_cast<int>(req.file_format), filter_expression_vector,
+      partition_expression_vector, dataset_schema_vector, projected_schema_vector);
+  builder.Finish(request);
+  uint8_t* buf = builder.GetBufferPointer();
+  int size = builder.GetSize();
+
+  bl.append((char*)buf, size);
+  return arrow::Status::OK();
+}
+
+arrow::Status DeserializeScanRequest(ScanRequest& req, ceph::bufferlist bl) {
+  auto request = flatbuf::GetScanRequest((uint8_t*)bl.c_str());
+
+  auto filter_expression_ = arrow::compute::Deserialize(
+                                std::make_shared<arrow::Buffer>(
+                                    request->filter()->data(), request->filter()->size()))
+                                .ValueOrDie();
+  req.filter_expression = filter_expression_;
+
+  auto partition_expression_ =
+      arrow::compute::Deserialize(
+          std::make_shared<arrow::Buffer>(request->partition()->data(),
+                                          request->partition()->size()))
+          .ValueOrDie();
+  req.partition_expression = partition_expression_;
+
+  arrow::ipc::DictionaryMemo empty_memo;
+  arrow::io::BufferReader projection_schema_reader(request->projection_schema()->data(),
+                                                   request->projection_schema()->size());
+  arrow::io::BufferReader dataset_schema_reader(request->dataset_schema()->data(),
+                                                request->dataset_schema()->size());
+
+  req.projection_schema =
+      arrow::ipc::ReadSchema(&projection_schema_reader, &empty_memo).ValueOrDie();
+  req.dataset_schema =
+      arrow::ipc::ReadSchema(&dataset_schema_reader, &empty_memo).ValueOrDie();
+
+  req.file_size = request->file_size();
+  req.file_format = (SkyhookFileType::type)request->file_format();
+  return arrow::Status::OK();
+}
+
+arrow::Status SerializeTable(std::shared_ptr<arrow::Table> table, ceph::bufferlist& bl) {
+  auto buffer_output_stream = arrow::io::BufferOutputStream::Create().ValueOrDie();
+
+  auto options = arrow::ipc::IpcWriteOptions::Defaults();
+  auto codec = arrow::Compression::LZ4_FRAME;
+
+  options.codec =
+      arrow::util::Codec::Create(codec, std::numeric_limits<int>::min()).ValueOrDie();
```

Comment:
`std::numeric_limits<int>::min()` should be the default.  Don't specify the second argument in case that changes someday.